libs/ojsonnet/Eval_jsonnet_subst.ml

```diff
@@ -0,0 +1,913 @@
+  | O (l, v, r) -> (
+      match v with
+      | Object (asserts, fields) ->
+          let new_asserts =
+            Common.map (fun (tk, expr) -> (tk, substitute id sub expr)) asserts
+          in
+          let new_fields =
+            Common.map
+              (fun { fld_name; fld_hidden; fld_value } ->
+                match fld_name with
+                | FExpr (l, name, r) ->
+                    {
+                      fld_name = FExpr (l, substitute id sub name, r);
+                      fld_hidden;
+                      fld_value = substitute id sub fld_value;
+                    })
+              fields
+          in
+          O (l, Object (new_asserts, new_fields), r)
+      | ObjectComp _ -> O (l, v, r) (* TODO *))
+  | Id (s, tk) -> if s =*= id then sub else Id (s, tk)
+  | IdSpecial id -> IdSpecial id
+  | Call
+      ( ArrayAccess
+          ( Id ("std", std_tok),
+            ( l1,
+              L (Str (None, DoubleQuote, (l2, [ (meth_str, meth_tk) ], r2))),
+              r1 ) ),
+        (l, args, r) ) ->
+      let new_args =
+        Common.map
+          (fun arg ->
+            match arg with
+            | Arg e -> Arg (substitute id sub e)
+            | NamedArg (ident, tk, e) ->
+                NamedArg (ident, tk, substitute id sub e))
+          args
+      in
+      let new_std = if is_imp_std meth_str then Id ("std", std_tok) else sub in
+      Call
+        ( ArrayAccess
+            ( new_std,
+              ( l1,
+                L (Str (None, DoubleQuote, (l2, [ (meth_str, meth_tk) ], r2))),
+                r1 ) ),
+          (l, new_args, r) )
+  | Local (_tlocal, binds, _tsemi, e) ->
+      if bind_list_contains binds id then Local (_tlocal, binds, _tsemi, e)
+      else
+        let new_binds =
+          Common.map
+            (fun (B (name, tk, expr)) -> B (name, tk, substitute id sub expr))
+            binds
+        in
+        Local (_tlocal, new_binds, _tsemi, substitute id sub e)
+  | ArrayAccess (v1, (l, v2, r)) ->
+      ArrayAccess (substitute id sub v1, (l, substitute id sub v2, r))
+  | Call (e0, (l, args, r)) ->
+      let new_func = substitute id sub e0 in
+      let new_args =
+        Common.map
+          (fun arg ->
+            match arg with
+            | Arg e -> Arg (substitute id sub e)
+            | NamedArg (ident, tk, e) ->
+                NamedArg (ident, tk, substitute id sub e))
+          args
+      in
+      Call (new_func, (l, new_args, r))
+  | UnaryOp ((op, tk), e) -> UnaryOp ((op, tk), substitute id sub e)
+  | BinaryOp (el, (op, tk), er) ->
+      BinaryOp (substitute id sub el, (op, tk), substitute id sub er)
+  | If (tif, e1, e2, e3) ->
+      If (tif, substitute id sub e1, substitute id sub e2, substitute id sub e3)
+  | Error (tk, e) -> Error (tk, substitute id sub e)
+  | ExprTodo ((_s, _tk), _ast_expr) -> ExprTodo ((_s, _tk), _ast_expr)
+
+(*TODO*)
+
+let rec substitute_kw kw sub expr =
```

Comment:
maybe add in comment reference to the spec, which part of the spec this function refers to or implement?

Reply:
So then one can look at the code and find the corresponding part in the spec and check if they match.